### PR TITLE
feat(discovery): Podman platform

### PR DESCRIPTION
**Pull request by andrewazores**
# Welcome to Cryostat! 👋
## Before contributing, make sure you have:
* [x] Read the [contributing guidelines](https://github.com/cryostatio/cryostat/blob/main/CONTRIBUTING.md)
* [x] Linked a relevant issue which this PR resolves
* [x] Linked any other relevant issues, PR's, or documentation, if any
* [x] Resolved all conflicts, if any
* [x] Rebased your branch PR on top of the latest upstream `main` branch
* [x] Attached at least one of the following labels to the PR: `[chore, ci, docs, feat, fix, test]`
* [x] Signed the last commit: `git commit --amend --signoff`
_______________________________________________

Related to https://github.com/cryostatio/cryostat/pull/492
Related to https://github.com/cryostatio/cryostat/issues/421

## Description of the change:
I revisited an old, old branch (June *2021* - unfortunately the original commit dates are lost now that this has been rebased and signed) where I was experimenting with talking to the Podman HTTP API from within Cryostat. At the time the Discovery tree querying work was under way, but there was not yet any notion of a persistent database for this data, and no developed system for discovery plugins other than the runtime detection of the single type platform strategy to pick up and use. At that time the original goal of getting a Podman HTTP API discovery query working was just to enable that discovery tree structure development for a contributor who was unable to run `crc` or other heftier systems that can provide a more interesting and non-flat deployment structure as compared to JDP's flat list.
Today we have much more sophistication around discovery, persistence, and modularity, and so I got the Podman HTTP API communication working and wrapped it up into a built-in discovery plugin.

## Motivation for the change:
This may be helpful because it is an additional example of a discovery plugin implementation, but it also removes dependency on JDP which may relieve some work from upstream JMC-core refactoring. JDP is probably only used in toy deployments of Cryostat like the upstream development `smoketest.sh`, so dropping it is unlikely to cause anyone any real issues so long as we have a reasonable development cycle replacement. Querying the Podman API periodically seems to do just as well for this purpose, and as an added bonus also is a little closer to other container runtimes in that we could configure multiple pods along with multiple containers within the pods and emulate something more like an OpenShift deployment (more complex topology).

In this PR the JDP platform client is still present, still named Default, and still has the lowest priority of the built-in plugins. The system will still choose the highest priority available built-in client to run, alongside Custom Targets and the plugin API, so in `smoketest.sh`/`run.sh` the Podman setup will be active rather than JDP. In integration tests the Podman socket is not active and mounted into the container, so the system continues to fall back on the old JDP discovery built-in.

## How to manually test:
1. *Run CRYOSTAT_IMAGE=quay.io... sh smoketest.sh...*
2. List targets, view topology if you have that PR checked out, etc.

### CODE REVIEW

**Comment by andrewazores** [comment]
Screenshot_2023-03-09_18-57-25

**Comment by andrewazores** [comment]
This + https://github.com/cryostatio/cryostat-web/pull/891 =

image

File: src/main/java/io/cryostat/net/HttpServer.java [hunk 1]
```diff
@@ -82,7 +82,10 @@ public class HttpServer extends AbstractVerticle {
                                         .setPort(netConf.getInternalWebServerPort())
                                         .addWebSocketSubProtocol("*")
                                         .setCompressionSupported(true)
-                                        .setLogActivity(true)));
+                                        .setLogActivity(true)
+                                        .setTcpFastOpen(true)
```
Comment (maxcao13) [inline src/main/java/io/cryostat/net/HttpServer.java:86]: Just curious what is reason for enabling these tcp features now?

**Comment by maxcao13** [comment]
image

Running the smoketest with the podman socket open results in both the Podman Realm and the JDP realm visible. Does this make sense?

**Comment by maxcao13** [comment]
Hmm... weird but it seems like jvmIds are different for two equivalent JVMs in the two realms. 
image
image

**Comment by andrewazores** [comment]
> image
> 
> Running the smoketest with the podman socket open results in both the Podman Realm and the JDP realm visible. Does this make sense?

Ah right, I forgot to mention. Since the built-in discovery only loads one of these plugins at runtime (plus custom targets, always), it's the Podman realm that is up to date. The JDP realm you're seeing is old data in the database from the last time you ran it and JDP was loaded.

I'm not sure what I want to do about that. Maybe the builtins that are not activated should have their realms deleted automatically?

**Comment by maxcao13** [comment]
Oh, that makes a lot of sense then. Then the jvmID thing is not an issue.

**Comment by andrewazores** [comment]
> I'm not sure what I want to do about that. Maybe the builtins that are not activated should have their realms deleted automatically?

External plugins that go through the API have the ping-pruning mechanism that will result in their realms being deleted if the plugins becomes unreachable however, so they do potentially suffer the same problem. And just in general, if a target disappears, that context is already lost for any archived recordings that came from it, which is why we have separate archived recording metadata that we store. So deleting the builtin plugins realms' doesn't seem too much different.

On a technical level it will take some refactoring, but I don't think it should be too hard to do.

**Comment by andrewazores** [comment]
Okay, last set of commits:

1. Allows multiple platform strategies to be selected. The highest priority available strategy continues to provide the sole AuthManager implementation. All selected strategies provide PlatformClients, which are what the internal implementation calls a built-in discovery plugin still. Custom Targets is special cased and always enabled. JDP happens to have its availability method hardcoded to returning `true`. The Podman client checks for the existence of the Podman API Unix socket. These may all be available and publishing to the discovery database at once, along with any external discovery plugins (like -agents). The previous variables for selecting a specific auth manager, a specific platform strategy, and for disabling built-ins (except custom targets) are all still there and work the same way.
2. Deregisters built-in plugins if they are unavailable on startup. For example, `CRYOSTAT_PLATFORM=io.cryostat.platform.internal.PodmanPlatformStrategy sh smoketest.sh` will cause the internal JDP built-in plugin to be deregistered and its discovery realm deleted from the tree. A subsequent rerun with no `CRYOSTAT_PLATFORM` setting will allow JDP to be selected alongside Podman again and it will re-register again as a new built-in discovery plugin.

**Comment by andrewazores** [comment]
Maybe `CRYOSTAT_PLATFORM` should support a comma-separated list instead of just a single platform now, too.

**Comment by maxcao13** [comment]
Running smoketest just like this: `sh smoketest.sh` runs both the podman client discovery platform and the JDP discovery platform. But running it with the env var `CRYOSTAT_PLATFORM=io.cryostat.platform.internal.PodmanPlatformStrategy sh smoketest.sh` removes the stale JDP built-in targets from the database and runs the podman discovery correctly on its own.

**Comment by andrewazores** [comment]
> Running smoketest just like this: `sh smoketest.sh` runs both the podman client discovery platform and the JDP discovery platform.
> 
> But running it with the env var `CRYOSTAT_PLATFORM=io.cryostat.platform.internal.PodmanPlatformStrategy sh smoketest.sh` removes the stale JDP built-in targets from the database and runs the podman discovery correctly on its own.

Yep, that sounds like it's working as expected.

**Review by maxcao13 — APPROVED** [review]
Looks good to me!

**Review by ebaron — DISMISSED** [review]
Dependency addition looks fine

**Pull request merged**